### ISSUE TITLE
Pointfree

### ISSUE THREAD

**Issue by samhh** [jonaskello/eslint-plugin-functional#154]
Would it be possible to lint for `x => f(x)` and recommend `f`? There are some circumstances in which these aren't equivalent but I think they don't apply if you're enforcing currying.

**Comment by RebeccaStevens** [comment]
There is a tslint rule that does this: [no-unnecessary-callback-wrapper](https://palantir.github.io/tslint/rules/no-unnecessary-callback-wrapper/).
I wasn't able to find an eslint version of the rule though when I did a quick google.

It's definitely possible for us to support this rule but we'd need to take care with how it's implemented.
We'd probably have to enforce that `x => f(x)` is only a violation if `f` is typed as taking no more parameters than the wrapper function being declared (so in this case 1).

Example of where `x => f(x)` cannot be simply replaced with `f`:

```ts
const f = (a = 0, b = 0) => a + b;

const foo = [1, 2, 3].map(x => f(x)); // [1, 2, 3]
const bar = [1, 2, 3].map(f);         // [1, 3, 5]

/*
 * The reason the outputs are different is because `.map` passes 3 arguments to the callback it is given.
 */
```

**Comment by RebeccaStevens** [comment]
I think if we were to add this, it would be our only stylist rule.

@jonaskello Would we want to add this rule?

**Comment by jonaskello** [comment]
I think the idea of this rule is nice and could potentially help people new to fp to understand that this kind of wrapping is unnecessary. I know I have written wrappers like that myself :-).

 I agree we must take care if we implement it. Here is a description of an [issue](https://medium.com/@nicolasdubien/beware-of-linters-41c82b075890) with the equivalent tslint rule. So like @RebeccaStevens said above we need to be carful with functions that takes variable arguments. I also searched for something like this in eslint and could not find much. The closest I found was [this rule](https://github.com/bahmutov/eslint-rules#potential-point-free) which is named `potential-point-free`. Perhaps having a softer rule name like this would communicate that it will not be correct in 100% of the cases. The best would of course be if we can be sure it will always be correct :-).

**Comment by RebeccaStevens** [comment]
I made a start on this implementing this rule. It seems to be handling the basic case right now.

**Comment by RebeccaStevens** [comment]
This is release in version 3.2.0 as an opt-in rule (i.e. it's not included in the recommended config).

Recommended way to opt-in:

```jsonc
{
  // ...
  "extends": [
    // ...
    "plugin:functional/stylitic"
  ],
  // ...
}
```

This rule will probably be added to the recommend config in the next major release.
